### ISSUE TITLE
Implement diff unit tests for libgit2

### ISSUE THREAD

**Issue by hausdorf** [basepi/libgit2#24]
Try to re-use as many use cases as possible from git diff unit tests.  Use the libgit2 test framework.

**Comment by trane** [comment]
Agreed.

**Comment by basepi** [comment]
Agreed.

**Comment by basepi** [comment]
Interestingly, all of git's unit tests are just shell scripts.  I guess that makes sense.

Since we're using a C linkable library, should our tests all be in C programs as opposed to shell scripts?

Also, there are a TON of tests for git.  Ridiculous.  They're in the `t/` directory, if you're interested in checking them out.

**Comment by basepi** [comment]
I just did `ls *diff*` to find all of the diff ones.

**Comment by basepi** [comment]
Lied again.  It uses shell scripts which create and then compile and then run and then destroy C programs.  Weird.

**Comment by basepi** [comment]
Lied one more time.  It creates c programs, then diffs them, then checks the output.  I was right the first time.  Ignore me for awhile, eh?

**Comment by hausdorf** [comment]
That makes sense as testing goes.

**Comment by basepi** [comment]
Can we do our testing via shell scripts?  I don't think so, since it's a linkable library.  However, assuming you guys have formatted your diffs EXACTLY like git does, then we could use their test cases at least.

**Comment by hausdorf** [comment]
What about making it a linkable library precludes us from using these tests?

**Comment by basepi** [comment]
Well, we don't have a command-line interface.  So you can't just run `git diff` from the command line, as these tests do.  You have to do it from within a C program that links to the libgit2 library, right?

**Comment by hausdorf** [comment]
So make a program with git's CLI.

**Comment by basepi** [comment]
Because that's not a huge task.  Definitely not a whole project on its own,
creating a minimal client that uses libgit2.

Sarcasm aside, I'll look into it.

**Comment by hausdorf** [comment]
We've implemented diff with no flags whatsoever. Realistically how much does that take to write up?

**Comment by basepi** [comment]
Right, but the tests rely on getting stuff out of a git repository, not just
diffing two files.  And I haven't looked, but I expect some (or many) of
them to use `commit` commands and so forth, to set up those snapshots for
the `diff` operations.  Additionally, I don't think we want to use shell
scripts, because the existing tests for libgit2 do not.  I'm inspecting
their testing framework now.  (Which seems to be written in C)

**Comment by hausdorf** [comment]
Guess we'll find out. But obviously the repo and commit stuff is mostly written. Just saying. Either way, their data and test cases are probably useful to us.

**Comment by basepi** [comment]
I'm definitely going to stick to libgit2's implemented testing framework.  Seems like it will be fairly straightforward (if time-consuming) to construct tests within that system.  I should be able to use the test cases from git's diff tests, however.  Again, assuming we format our diff output exactly the same.

I'm going to bed for now, will continue this tomorrow morning before church and whatnot.

**Comment by hausdorf** [comment]
Your choice for sure man, but the soonest you can get this done truly is the best. I _really_ could use it.

**Comment by basepi** [comment]
I understand.  Urgent enough that I should take the fastest route to usable tests, even if we'll have to write new ones for the pull request later?  (i.e. should I sacrifice their testing framework if it gets you tests sooner?)

**Comment by basepi** [comment]
Obviously either way I'm going to shoot to be done by Tuesday.

**Comment by hausdorf** [comment]
No, we still have time. I can administer crude hand-tests myself.

**Comment by basepi** [comment]
Cool.

**Comment by hausdorf** [comment]
Well, it would really be nice if they could be done before we actually show the presentation.

**Comment by basepi** [comment]
Right.  Oh and btw:

>  I can administer crude hand-tests myself.

<Insert generic masturbation joke here>

**Comment by hausdorf** [comment]
As in, tomorrow. Maybe Monday. That would be really awesome, if you can manage.

**Comment by hausdorf** [comment]
And of course, by "Monday" I mean tomorrow.*

**Comment by hausdorf** [comment]
*where "tomorrow" means in 5 minutes, which is when I will need to use them.

**Comment by trane** [comment]
Wow, I've been cleaning and organizing for the entire evening. I open my computer up and I have a novelette to read. Thanks gents. @DrSleep @crakdmirror @kyeana <- that's a first for me. I have never ever ever ever ever ever ever........

**Comment by basepi** [comment]
Updated the issue title/text.

I'll tell you what, though:  The test framework is as confusing as hell.  There are not nearly enough `#include`s.......

Gotta figure out how to compile/run the test framework as well.  I'll keep chugging.

**Comment by basepi** [comment]
@DrSleep, if we can't get prettified output from our functions, then I need example cases of what the output looks like so I can actually _check_ the output.  I suppose I could just run the functions and then plug that output into the tests, but that's _not_ good testing procedure.  Git has many use cases that we could leverage, but we'd have to bring our output in line _exactly_ with their Git's output.

**Comment by hausdorf** [comment]
I am aware of the issue. I will get something together when I can.

**Comment by basepi** [comment]
No worries.  I don't think we actually _need_ unit tests in the submitted code for 3505 (though they would be nice), but we'll need them to get our pull request accepted by libgit2 later.

**Comment by hausdorf** [comment]
Formally, my goal is to build something that we can actually turn in. I think that's reasonable.

**Comment by hausdorf** [comment]
Incidentally, I'm going to be working on this most of tomorrow, and will probably need you to help me review for algorithms as a result.

**Comment by basepi** [comment]
Algorithms.....I'm kinda worried about that one.  I know Jeremiah Turner and some other people are getting together to study tomorrow like noon-ish.  I may go study with them.  Also would be happy to study with you, just can't in the evening.  (That's when this whole church road-show thing is going down)  Incidentally, this means that anything I need to do for the code freeze needs to be done before tomorrow evening.  Though the road show stuff will probably be done by like 2100-2200, so I'll be able to help for the last couple of hours.  (Assuming it's due at midnight, which I have not verified)

**Comment by basepi** [comment]
The code freeze is 2340 tomorrow.  According to the assignment page.

Need to include a usage README in there, shouldn't be a problem, the default README.md is probably enough, and we can modify it slightly for our group's information.
